### ISSUE TITLE
HTML_CodeSniffer outdated - wrongfully not reporting errors

### ISSUE THREAD

**Issue by MichielDHVD** [dso-toolkit/dso-toolkit#88]
De wcag checker geeft geen errors weer. 

Hier zijn twee instanties waar de checker geen errors aan geeft terwijl deze er wel zijn. Ik gebruik een Html-codesniffer die zowel errors als warnings geeft. 

Navs (Onder het kapotje navigatie) 

_This element's role is "presentation" but contains child elements with semantic meaning._
Alle 6 de elementen hebben deze error. 

Buttons. 

De checker geeft aan dat het contrast niveau van de witte tekst en oranje achtergrond kleur niet groot genoeg is. De 18px bold font text is net te klein om het contrast niveau te halen. Deze moet opgeschaald worden naar 20px.

_This element has insufficient contrast at this conformance level. Expected a contrast ratio of at least 4.5:1, but text in this element has a contrast ratio of 4.04:1. Recommendation: change background to #c75300._

**Comment by VolleMelk** [comment]
In welke browser met welke checker werk jij?

**Comment by MichielDHVD** [comment]
In chrome met Html_CodeSniffer http://squizlabs.github.io/HTML_CodeSniffer/

**Comment by VolleMelk** [comment]
De HTML_CodeSniffer die vanuit een bookmarklet werkt, gebruikt altijd de laatste versie.

De buildstraat van de DSO Toolkit maakt gebruik van gulp-accessibility -> access_sniff -> HTML_CodeSniffer.

In dit "treintje" is een achterstand ontstaan. Ik heb de packages geupdate, waardoor onze resultaten gelijk kwamen.

Vervolgens bij alle componenten behalve badge de WCAG errors opgelost. Badge wordt in #125 opgelost en heb ik bewust overgeslagen.

Ondertussen zijn de navs aangepast en is het kleurenschema anders geworden. Jouw contrast melding en oplossing van 18px naar 20px is niet meer van toepassing.

**Comment by MichielDHVD** [comment]
@VolleMelk Is er een specifiek overzicht van de aanpassingen. Want als er kleuren zijn aangepast welke kleuren zijn dat dan ? Ik heb iets meer informatie nodig betreffende de aanpassingen. Of desnoods de issues waar het allemaal is toegepast.

**Comment by VolleMelk** [comment]
Ik heb niet goed gelezen! De increment van 18px naar 20px gaat over de _buttons_, ik dacht over _navs_. En aangezien ik geen WCAG errors kreeg bij navs (of eigenlijk nergens behalve bij badges) dacht ik dat het was opgelost omdat in #87 wordt het navigatie component opgepakt ik dacht dat er al een merge was geweest waardoor de kleuren waren aangepast.

Opeenstapeling van fouten ;-)

Maar in ieder geval voorwaarts: Wij gaan de 18px opschroeven naar de 20px voor alle buttons.

Dan over de WCAG checker. De WCAG checker die wij (jij en ik) vanuit de browser draaien is HTML_CodeSniffer van Squiz. De bookmarklet is een loader die de code van HTML_CodeSniffer bootstrapt en het DOM scant.

De WCAG checker in de buildstraat is dezelfde WCAG checker, maar die draait niet vanuit de context van een browser. Het lukt de WCAG checker niet om de uiteindelijk gerenderde kleuren tegen elkaar aan te leggen. De WCAG ziet de problemen wel aankomen en geeft de melding als WARNING:

```
WARNING: WCAG2AA.Principle1.Guideline1_4.1_4_3.G18.BgImage
This element's text is placed on a background image. Ensure the contrast ratio between the text and all covered parts of the image are at least 4.5:1.
```

Ik kan nog 'n keer kijken of er een andere mogelijkheid is. De HTML_CodeSniffer bookmarklet doet z'n werk wel erg goed, ook omdat hij de kracht van een browser heeft.

**Comment by VolleMelk** [comment]
@MichielDHVD Ik kan de melding niet reproduceren op de buttons pagina: https://www.dso-toolkit.nl/2.0.0-alpha.1/components/preview/buttons.html

Dit lukt mij alleen als ik de HTML_CodeSniffer op AAA zet, maar ik dacht dat wij tot AA gingen.

**Comment by MichielDHVD** [comment]
Hmm dat is inderdaad vreemd. Bij mij pikt het ook pas bij AAA op. 

Maar in de guidelines zelf staat toch echt het volgende. 

WCAG 2.0 level AA requires a contrast ratio of at least 4.5:1 for normal text and 3:1 for large text. Level AAA requires a contrast ratio of at least 7:1 for normal text and 4.5:1 for large text.

Dus of contrast van kleur moet 4.5:1 zijn tussen tekst met de grootte van 18.66 of kleiner. 

3:1 voor tekst groter dan 18.66 

Er is een website waar je dit heel makkelijk kunt checken. 

https://webaim.org/resources/contrastchecker/

Als je hier aan de linker kant de kleur #e17000 ingeeft dan kun je zien dat alleen Large text de AA haalt. 
Dus ik weet niet waarom die codesniffer hem nu niet weergeeft. Enigste optie zou zijn dat hij de tekst als large text leest. 

Maar het klopt dat we niet to AAA gaan (Dat is praktisch gezien onmogelijk om te halen te veel variabelen)

**Comment by VolleMelk** [comment]
BREAKING zijn changes waar de ontwikkelaar zijn bestaande codebase voor moet nalopen om actueel met de voorschriften te blijven.

Bij deze issue is de volgende change doorgevoerd: https://github.com/dso-toolkit/dso-toolkit/pull/98/files#diff-3be5944bf634cf23cae31e2ff47b5b93

**Comment by VolleMelk** [comment]
Ik begin steeds meer te twijfelen aan de WCAG implementatie dmv. de HTML_CodeSniffer in de buildstraat. De resultaten vallen tegen:
* Een aantal regels worden niet gecheckt, de CodeSniffer kan zonder een browser geen contrast bepalen en geeft een lege melding
* Veel dezelfde generieke meldingen (Check that the document has a valid and meaningful TITLE)
* Afwijkingen ten opzichte van de specificaties
* Afwijkingen ten opzichte van de HTML_CodeSniffer bookmarklet.

Hoe kijken jullie tegen de implementatie aan?

**Comment by MichielDHVD** [comment]
We hebben besproken dat we dit systeem op een andere manier moeten gaan aanpakken. 

Dus de labels en tekstuele elementen worden gechecked door het huidige systeem. Maar voor contrast checken moeten we een andere oplossing vinden. 

Dat is kort samengevat wat we besproken hadden tijdens de meeting in Rotterdam. 

Enigste wat nu moet gebeuren is dus iets vinden dat de elementen kan opvangen die deze checker niet kan controleren. 

Moet hier een apart issue voor gemaakt worden?

**Comment by tfrijsewijk** [comment]
Ja, laten we dat doen. Dan kan deze gesloten worden
